src/VisualStudio/CSharp/Impl/ProjectSystemShim/CSharpProjectShim.ICSharpProjectSiteEx.cs

```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.CSharp.ProjectSystemShim
+{
+    internal partial class CSharpProjectShim : ICSharpProjectSiteEx
+    {
+        public void OnSourceFileAddedEx(string filename, string folder)
```

Comment:
Is there an agreement here that synchronization is being done by the project system to avoid parallel calls?

Reply:
(As it stands, your AddFile isn't mutating any state and VisualStudioProject is fine itself...but I'm not sure if that'll ever hold if somebody else adds something further to AddFile.)

Reply:
Yes, it's on the project system to deal with that.

Ideally these interfaces are just a temporary stepping stone and I'll eventually be able to remove them in favor of the legacy project system shims implementing `IWorkspaceProjectContext`.